### PR TITLE
Merging SCEs: Use union of features

### PR DESCRIPTION
**Pull request by sjspielman**
Towards #150 

This PR takes some next steps towards simplifying altexp merging:

- A little reorganization in `merge_sce_list()` so that all checks that might cause us to `stop()` are at the top of the function, for faster failing.
- Add a check for altExps: Each given altexp name must have the same features, or we `stop()`.
- Create new function `create_sce_with_all_features()`, which is currently only used for the main SCE (altExp usage in next PR), which ensures that a given vector of features (here, vector representing union of relevant features) is always reflected in the SCE. I am extra careful to ensure the row order matches, as well. 
  - There is one `TODO` here about how to handle rowData columns like `gene_ids` which may get filled with `NA`s. I haven't really explored this situation in depth; waiting to do so in the next PR(s) when I start applying this function to the altExps where, at least for our current ScPCA data, this is likely to matter. 
  - If you have specific thoughts though, happy to get going on this TODO now!
- Tests are present via _actual test-driven development_

### CODE REVIEW

**Comment by jashapiro** [comment]
> * Create new function `create_sce_with_all_features()`, which is currently only used for the main SCE (altExp usage in next PR), which ensures that a given vector of features (here, vector representing union of relevant features) is always reflected in the SCE. I am extra careful to ensure the row order matches, as well.

I actually don't think we want to do this for main Exps. Sorry if I was not clear about that! There are some stats that could break with NA values, and I think it is better for us to leave the main experiments as they are, the intersection of the sampled genes.

We can keep this function around, but I don't think we want it to be used right now. For the AltExps, there may be an argument to use it, but I think what might actually be better to create separate AltExps if there are two with the same name that have different feature complements, e.g., `adt-1` and `adt-2`.

**Comment by jashapiro** [comment]
> > * Create new function `create_sce_with_all_features()`, which is currently only used for the main SCE (altExp usage in next PR), which ensures that a given vector of features (here, vector representing union of relevant features) is always reflected in the SCE. I am extra careful to ensure the row order matches, as well.
> 
> I actually don't think we want to do this for main Exps. Sorry if I was not clear about that! There are some stats that could break with NA values, and I think it is better for us to leave the main experiments as they are, the intersection of the sampled genes.
> 
> We can keep this function around, but I don't think we want it to be used right now. For the AltExps, there may be an argument to use it, but I think what might actually be better to create separate AltExps if there are two with the same name that have different feature complements, e.g., `adt-1` and `adt-2`.

Just following up on this a bit... having started to look at the code: I think you can leave most of what you have in the `prepare` function: we just won't give it the union by default...

**Review by jashapiro — COMMENTED** [review]
I made a number of comments below, but I am sort of undecided about how much of the "expansion" capabilities we want to keep. Part of me thinks we don't need it yet, as per my earlier comment I don't think we want to expand main SCE assays at this point. We may want to do that for AltExps, but for now I think merging only if the AltExps match is sufficient (we could allow feature order changes, if we want to be generous). 

But I don't want to abandon the work you have put in here, as it might be useful in the future! So the comments below are about what might happen if we wanted to keep this capability:

For the assays themselves, this looks fine, with some small modifications to keep things sparse.

But I think the rowData expansion needs to be rewritten, as the current codes doesn't handle mixed data types correctly. I would expect a separate function that takes a data frame and a list of features is probably going to be useful here: Basically much like the `create_sce_with_features` but `create_rowData_with_features` (I deleted `all`) to handle the bookkeeping there, as I think it will be a bit complicated and worth testing directly.

----

I think the best plan might be to branch off this code and link the branch in a new issue: "Allow SCE merge with unions of features" that we can come back to later, then go back to finishing the main feature we want for now, which would leave the main object merge as it is and add the ability to include AltExps if they are all the same type (or missing).

File: R/merge_sce_list.R [hunk 1]
```diff
@@ -97,6 +95,64 @@ merge_sce_list <- function(
...
+    }
+  }
+
+  ## Subset SCEs to shared features and ensure appropriate naming ------------------
+
+  # First, obtain the union of features for all (main) SCE objects
+  # this will also be the final order of features/rows in the final SCE
+  sce_full_features <- sce_list |>
+    purrr::map(rownames) |>
+    purrr::reduce(union)
```
[the remark this answers is not in the document]
Comment (sjspielman) [reply R/merge_sce_list.R:154]: Noting this `intersect` code is already higher up in the script as a _check_ to ensure there are intersecting features (I moved it there in this PR), so I'll just delete this and use what we've got above.

**Comment by sjspielman** [comment]
After some additional discussion, we'll circle back to this later with #249.

Note that some of these commits will probably be cherry-picked out into other PRs to address #150.

**Comment by sjspielman** [comment]
We're _very_ unlikely to proceed with this code, so I'm going to close this PR out. But the branch lives on and is linked from #249 juuust in case.

**Pull request closed without merging**